src/monitor/node_active_protocol.c

```diff
@@ -221,16 +221,20 @@ register_node(PG_FUNCTION_ARGS)
 	}
 
 	/*
-	 * When adding a second node to a formation that has number_sync_standbys
-	 * set to zero (the default value for single node and single standby
-	 * formations), we switch the default value to 1 automatically.
+	 * When adding a second sync node to a formation that has
+	 * number_sync_standbys set to zero (the default value for single node and
+	 * single standby formations), we switch the default value to 1
+	 * automatically.
 	 */
-	if (formation->number_sync_standbys == 0)
+	if (pgAutoFailoverNode->goalState == REPLICATION_STATE_WAIT_STANDBY &&
+		formation->number_sync_standbys == 0)
 	{
-		List *allNodes = AllAutoFailoverNodes(formationId);
+		AutoFailoverNode *primaryNode =
+			GetPrimaryNodeInGroup(formationId, currentNodeState.groupId);
+		List *standbyNodesList = AutoFailoverOtherNodesList(primaryNode);
+		int standbyNodeCount = CountSyncStandbys(standbyNodesList);
 
-		/* second standby to be registered */
-		if (list_length(allNodes) == 3)
+		if (standbyNodeCount == 2)
```

Comment:
Good point, I added the following comment:

``` c
		/*
		 * number_sync_standbys = 0 is a special case in our FSM, because we
		 * have special handling of a missing standby then, switching to
		 * wait_primary to disable synchronous replication when the standby is
		 * not available.
		 *
		 * For other values (N) of number_sync_standbys, we require N+1 known
		 * sync standby nodes, so that you can lose a standby at any point in
		 * time and still accept writes.
		 *
		 * The default value for number_sync_standbys with two standby nodes is
		 * 1. Because it was set to zero when adding the first standby, we need
		 * to increment the value when adding a second standby node that
		 * participates in the replication quorum (a "sync standby" node).
		 */
```